[MPS] Make fused rms_norm traceable

malfet opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #150661

Which is a regression, introduced by https://github.com/pytorch/pytorch/issues/150629#issue-2970312779 which I should have reviewed more thoroughly.

- Defined `_fused_rms_norm`, added MPS-only implementation for it and dispatch from `rms_norm_symint`,  which is registered as `CompositeImplicitAutograd`, i.e. it is not supposed to do any computations over Tensor, only dispatch to other ops
- 
- Register `_fused_rms_norm` as a fallback in `torch/_inductor/lowering.py`
- Added unit test to avoid those regressions in the future

TODO:
- Get rid of this op, change `rms_norm_symint` definition to `CompositeExplicitAutograd` and implement backward function in `tools/autograd/derivatives.yaml`
- Benchmark compiler and re-enable decomp as follows when compiled code is faster
```python
@register_decomposition(aten._rms_norm_fused)
def rms_norm_fused(
    self: torch.Tensor, ndim: int, weight: torch.Tensor, eps: float
) -> torch.Tensor:
    dtr = [self.dim() - i - 1 for i in range(ndim)]
    return self * weight * (self.pow(2).mean(dtr, keepdim=True).add(eps).rsqrt())
```



Fixes https://github.com/pytorch/pytorch/issues/150629


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @ipiszy @chenyang78 @kadeng @muchulee8 @amjames @chauhang @aakhundov

malfet committed:
[MPS] Make fused rmsnorm traceable

By declaring it as an ATen op

Fixes https://github.com/pytorch/pytorch/issues/150629

[ghstack-poisoned]

malfet committed:
Update on "[MPS] Make fused rms_norm traceable"


By declaring it as an ATen op

Fixes https://github.com/pytorch/pytorch/issues/150629

[ghstack-poisoned]

kimishpatel commented:
Can you add in summary why #150629 introduced a regregssion? Also do we know if inductor generated code will be more performant compared to native kernel?

malfet committed:
Update on "[MPS] Make fused rms_norm traceable"


Which is a regression, introduced by https://github.com/pytorch/pytorch/issues/150629#issue-2970312779 which I should have reviewed more thoroughly.

- Defined `_rms_norm_fused`, added MPS-only implementation for it and dispatch from native::rms_norm_symint there in no-grad mode
- Defined a decomp for it in `torch/_inductor/decomposition.py`
- Added unit test to avoid those regressions in the future

TODO/Ideas:
 - Perhaps define it as non-decomposable
 - Make `torch.compiler.is_compiling` reflect to some sort of  `at::Context` property

Fixes https://github.com/pytorch/pytorch/issues/150629

cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

manuelcandales commented on torch/_inductor/decomposition.py:
```diff
@@ -1105,3 +1105,11 @@ def rrelu_with_noise_functional(
         return aten.leaky_relu(self, negative_slope), torch.Tensor()
+
+
+@register_decomposition(aten._rms_norm_fused)
```
We don't want this to happen. We want inductor to use the fused implementation, not to decompose it.

albanD commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3302,6 +3302,10 @@
   dispatch:
     CompositeImplicitAutograd: rms_norm_symint
 
+- func: _rms_norm_fused(Tensor input, int normalized_shape_ndim, Tensor weight, float eps) -> Tensor
```
tbh I think you're making your life more complex for no good reason here.
You can set a MPS dispatch on rms_norm above, define a backward formula for it and call it a day.

malfet committed:
Update

[ghstack-poisoned]

manuelcandales commented on torch/_inductor/lowering.py:
```diff
@@ -2625,6 +2625,7 @@ def is_aligned(x):
 make_fallback(aten.exponential.default, warn=False)  # (fails accuracy on test_torch.py)
 make_fallback(aten._pdist_forward)  # Has decomp. Needs benchmarks
 make_fallback(aten.soft_margin_loss_backward, warn=False)  # py_impl?
+make_fallback(aten_rms_norm_fused, warn=False)  # (MPS-only and faster than decomp)
```
```suggestion
make_fallback(aten._rms_norm_fused, warn=False)  # (MPS-only and faster than decomp)
```

manuelcandales commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -3301,6 +3301,10 @@
   dispatch:
     CompositeImplicitAutograd: rms_norm_symint
 
+- func: _rms_norm_fused(Tensor input, int normalized_shape_ndim, Tensor weight, float eps) -> Tensor
```
I am not sure if there is some naming convention, or if it's just coincidence, but when I search the word "fused" in native_functions.yaml, it is always before the name of the function. If there is such naming convention, then this should be changed to `_fused_rms_norm`

malfet committed:
Update

[ghstack-poisoned]

malfet committed:
Update

[ghstack-poisoned]

malfet committed:
Update

[ghstack-poisoned]

malfet committed:
Update on "[MPS] Make fused rms_norm traceable"


Which is a regression, introduced by https://github.com/pytorch/pytorch/issues/150629#issue-2970312779 which I should have reviewed more thoroughly.

- Defined `_fused_rms_norm`, added MPS-only implementation for it and dispatch from `rms_norm_symint`,  which is registered as `CompositeImplicitAutograd`, i.e. it is not supposed to do any computations over Tensor, only dispatch to other ops
- 
- Register `_fused_rms_norm` as a fallback in `torch/_inductor/lowering.py`
- Added unit test to avoid those regressions in the future

TODO:
- Get rid of this op, change `rms_norm_symint` definition to `CompositeExplicitAutograd` and implement backward function in `tools/autograd/derivatives.yaml`
- Benchmark compiler and re-enable decomp as follows when compiled code is faster
```python
register_decomposition(aten._rms_norm_fused)
def rms_norm_fused(
    self: torch.Tensor, ndim: int, weight: torch.Tensor, eps: float
) -> torch.Tensor:
    dtr = [self.dim() - i - 1 for i in range(ndim)]
    return self * weight * (self.pow(2).mean(dtr, keepdim=True).add(eps).rsqrt())
```



Fixes https://github.com/pytorch/pytorch/issues/150629


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

malfet commented:
@pytorchbot merge -f "let's test in prod"

malfet commented:
@pytorchbot revert -m *Has decomp started to fail again" -c nosignal

malfet committed:
Update on "[MPS] Make fused rms_norm traceable"


Which is a regression, introduced by https://github.com/pytorch/pytorch/issues/150629#issue-2970312779 which I should have reviewed more thoroughly.

- Defined `_fused_rms_norm`, added MPS-only implementation for it and dispatch from `rms_norm_symint`,  which is registered as `CompositeImplicitAutograd`, i.e. it is not supposed to do any computations over Tensor, only dispatch to other ops
- 
- Register `_fused_rms_norm` as a fallback in `torch/_inductor/lowering.py`
- Added unit test to avoid those regressions in the future

TODO:
- Get rid of this op, change `rms_norm_symint` definition to `CompositeExplicitAutograd` and implement backward function in `tools/autograd/derivatives.yaml`
- Benchmark compiler and re-enable decomp as follows when compiled code is faster
```python
register_decomposition(aten._rms_norm_fused)
def rms_norm_fused(
    self: torch.Tensor, ndim: int, weight: torch.Tensor, eps: float
) -> torch.Tensor:
    dtr = [self.dim() - i - 1 for i in range(ndim)]
    return self * weight * (self.pow(2).mean(dtr, keepdim=True).add(eps).rsqrt())
```



Fixes https://github.com/pytorch/pytorch/issues/150629


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx ipiszy chenyang78 kadeng muchulee8 amjames chauhang aakhundov

[ghstack-poisoned]

malfet commented:
@pytorchbot merge -f "Re-added has decomposition"
